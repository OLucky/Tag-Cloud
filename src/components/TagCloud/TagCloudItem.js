import React from "react";
import {Link} from "react-router-dom";
import PropTypes from "prop-types";

const TagCloudItem = ({label, id, fontSize, color}) => (
	<Link className="tag-cloud_content_item" to={`/${encodeURIComponent(id)}`}>
		<span style={{fontSize, color}}>{label}</span>
	</Link>
);

TagCloudItem.propTypes = {
	label: PropTypes.string.isRequired,
	id: PropTypes.string.isRequired,
	fontSize: PropTypes.number.isRequired,
	color: PropTypes.string,
};

export default TagCloudItem;