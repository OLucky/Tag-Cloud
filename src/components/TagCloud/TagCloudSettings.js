import React, {Component} from "react";
import PropTypes from "prop-types";

class TagCloudSettings extends Component {

	settingChangeHandler = (event, label) => {
		const {onSettingChange} = this.props;
		const value = event.target.value;

		return onSettingChange({label, value});
	}

	colorChangeHandler = (event, index) => {
		const {onColorChange} = this.props;
		const color = event.target.value;
		
		return onColorChange(color, index);
	}

	displayColors = colors => {
		const {settings, onColorRemove} = this.props;
		const {numberOfPages} = settings;

		return colors.map((color, index) => (
			<div key={color + index}>
				<input onChange={event => this.colorChangeHandler(event, index)} type="color" value={color} />
				<button className="tag-cloud_settings_remove-color-btn" disabled={colors.length <= numberOfPages} onClick={() => onColorRemove(index)}>Remove Color</button>
			</div>
		));
	}

	render() {
		const {settings, onColorAdd} = this.props;
		const {minFontSize, maxFontSize, colors} = settings;

		return(
			<div className="tag-cloud_settings">
				<h2>Tag Cloud Settings</h2>
				<div className="tag-cloud_settings_item">
					<p>Minimum font size</p>
					<input type="number" onChange={event => this.settingChangeHandler(event, "minFontSize")} value={minFontSize} min="1" max={maxFontSize}/>
				</div>
				<div className="tag-cloud_settings_item">
					<p>Maximum font size</p>
					<input type="number" onChange={event => this.settingChangeHandler(event, "maxFontSize")}  value={maxFontSize} min={minFontSize}/>
				</div>
				{this.displayColors(colors)}
				<button className="tag-cloud_settings_add-color-btn" onClick={onColorAdd}>Add Color</button>
			</div>
		);
	}
}

TagCloudSettings.propTypes = {
	settings: PropTypes.object.isRequired,
	onColorAdd: PropTypes.func.isRequired,
	onColorChange: PropTypes.func.isRequired,
	onColorRemove: PropTypes.func.isRequired,
	onSettingChange: PropTypes.func.isRequired
};

export default TagCloudSettings;