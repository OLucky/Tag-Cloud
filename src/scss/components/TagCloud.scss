.tag-cloud_content {
	display: flex;
	flex-wrap: wrap;
	align-items: center;
}

.tag-cloud_content_item {
	display: block;
	margin: 0 1rem;
}