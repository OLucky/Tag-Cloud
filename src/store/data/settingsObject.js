export const settingsObject = {
	minFontSize: 16,
	maxFontSize: 160,
	colors: ["#000000", "#0000FF", "#FF00FF",
		"#008000", "#800000",
		"#000080", "#808000", "#800080",
		"#C0C0C0", "#008080", "#FF6347"],
	numberOfPages: 9
};