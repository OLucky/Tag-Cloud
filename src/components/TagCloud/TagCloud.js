import React, {Component} from "react";
import PropTypes from "prop-types";

import TagCloudItem from "./TagCloudItem";
import TagCloudSettings from "./TagCloudSettings";

class TagCloud extends Component {

	generateItems = (tags, settings) => {
		const maxScore = Math.max.apply(Math, tags.map(item => item.sentimentScore));
		const minScore = Math.min.apply(Math, tags.map(item => item.sentimentScore));
		const {colors} = settings;

		const minFontSize = parseInt(settings.minFontSize);
		const maxFontSize = parseInt(settings.maxFontSize);

		return tags.map(item => {
			const tagScore = item.sentimentScore;
			const tagFontSize = tagScore === minScore ? minFontSize : Math.floor((tagScore / maxScore) * (maxFontSize - minFontSize) + minFontSize);
			const color = colors[Math.floor(Math.random() * colors.length) - 1];

			return (<TagCloudItem label={item.label} id={item.id} key={item.id} fontSize={tagFontSize} color={color} />);
		});
	}

	render() {
		const {tags, settings, onSettingChange, onColorChange, onColorAdd, onColorRemove} = this.props;

		return (
			<div className="container tag-cloud">
				<h1 className="tag-cloud_title">
					Tag Cloud
				</h1>
				<div className="tag-cloud_content">
					{this.generateItems(tags, settings)}
				</div>
				<TagCloudSettings settings={settings} onSettingChange={onSettingChange} onColorChange={onColorChange} onColorAdd={onColorAdd} onColorRemove={onColorRemove} />
			</div>
		);
	}
}

TagCloud.propTypes = {
	tags: PropTypes.array.isRequired,
	settings: PropTypes.object.isRequired,
	onSettingChange: PropTypes.func.isRequired,
	onColorChange: PropTypes.func.isRequired,
	onColorAdd: PropTypes.func.isRequired,
	onColorRemove: PropTypes.func.isRequired,
};

export default TagCloud;