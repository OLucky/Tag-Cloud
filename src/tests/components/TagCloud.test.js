import React from "react";
import {render, shallow} from "enzyme";
import toJson from "enzyme-to-json";
import { StaticRouter } from "react-router";

import "../config";

import TagCloud from "../../components/TagCloud/TagCloud";
import TagCloudSettings from "../../components/TagCloud/TagCloudSettings";

import {tagsArray} from "../../store/data/tagsArray";
import {settingsObject} from "../../store/data/settingsObject";

const oneTag = tagsArray[0];

const mockData = [{
	label: oneTag.label,
	id: oneTag.id,
	sentimentScore: oneTag.sentimentScore
}];

global.Math.random = () => 0.5;
const context = {};

describe("TagCloud component", () => {

	test("renders correctly with children", () => {
		const onSettingChange = jest.fn();
		const onColorChange = jest.fn();
		const onColorAdd = jest.fn();
		const onColorRemove = jest.fn();

		const component = render(
			<StaticRouter context={context}>
				<TagCloud tags={mockData} settings={settingsObject} 
					onSettingChange={onSettingChange} onColorChange={onColorChange} onColorAdd={onColorAdd} onColorRemove={onColorRemove}/>
			</StaticRouter>);

		expect(toJson(component)).toMatchSnapshot();
	});

	describe("TestCloudSettings component", () => {

		test("calls setting actions", () => {
			const onSettingChange = jest.fn();
			const onColorChange = jest.fn();
			const onColorAdd = jest.fn();
			const onColorRemove = jest.fn();

			const component = shallow(<TagCloudSettings settings={settingsObject}
				onSettingChange={onSettingChange} onColorChange={onColorChange} onColorAdd={onColorAdd} onColorRemove={onColorRemove}/>);

			component.find("[type='number']").at(0).simulate("change", {target: {value: 1}});
			expect(onSettingChange.mock.calls.length).toBe(1);	
			expect(onSettingChange.mock.calls[0]).toEqual([{label: "minFontSize", value: 1}]);
		});

		test("calls color actions", () => {
			const onSettingChange = jest.fn();
			const onColorChange = jest.fn();
			const onColorAdd = jest.fn();
			const onColorRemove = jest.fn();

			const component = shallow(<TagCloudSettings settings={settingsObject}
				onSettingChange={onSettingChange} onColorChange={onColorChange} onColorAdd={onColorAdd} onColorRemove={onColorRemove}/>);

			component.find("[type='color']").at(0).simulate("change", {target: {value: "#FFFFFF"}});
			expect(onColorChange.mock.calls.length).toBe(1);	
			expect(onColorChange.mock.calls[0]).toEqual(["#FFFFFF", 0]);	

			component.find(".tag-cloud_settings_add-color-btn").simulate("click");
			expect(onColorAdd.mock.calls.length).toBe(1);

			component.find(".tag-cloud_settings_remove-color-btn").at(0).simulate("click");
			expect(onColorRemove.mock.calls.length).toBe(1);
			expect(onColorRemove.mock.calls[0]).toEqual([0]);	
		});
	});

});
