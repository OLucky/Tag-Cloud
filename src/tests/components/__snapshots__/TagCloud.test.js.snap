// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagCloud component renders correctly with children 1`] = `
<div
  class="container tag-cloud"
>
  <h1
    class="tag-cloud_title"
  >
    Tag Cloud
  </h1>
  <div
    class="tag-cloud_content"
  >
    <a
      class="tag-cloud_content_item"
      href="/1751295897__Odessa"
    >
      <span
        style="font-size:16px;color:#800000"
      >
        Odessa
      </span>
    </a>
  </div>
  <div
    class="tag-cloud_settings"
  >
    <h2>
      Tag Cloud Settings
    </h2>
    <div
      class="tag-cloud_settings_item"
    >
      <p>
        Minimum font size
      </p>
      <input
        max="160"
        min="1"
        type="number"
        value="16"
      />
    </div>
    <div
      class="tag-cloud_settings_item"
    >
      <p>
        Maximum font size
      </p>
      <input
        min="16"
        type="number"
        value="160"
      />
    </div>
    <div>
      <input
        type="color"
        value="#000000"
      />
      <button
        class="tag-cloud_settings_remove-color-btn"
      >
        Remove Color
      </button>
    </div>
    <div>
      <input
        type="color"
        value="#0000FF"
      />
      <button
        class="tag-cloud_settings_remove-color-btn"
      >
        Remove Color
      </button>
    </div>
    <div>
      <input
        type="color"
        value="#FF00FF"
      />
      <button
        class="tag-cloud_settings_remove-color-btn"
      >
        Remove Color
      </button>
    </div>
    <div>
      <input
        type="color"
        value="#008000"
      />
      <button
        class="tag-cloud_settings_remove-color-btn"
      >
        Remove Color
      </button>
    </div>
    <div>
      <input
        type="color"
        value="#800000"
      />
      <button
        class="tag-cloud_settings_remove-color-btn"
      >
        Remove Color
      </button>
    </div>
    <div>
      <input
        type="color"
        value="#000080"
      />
      <button
        class="tag-cloud_settings_remove-color-btn"
      >
        Remove Color
      </button>
    </div>
    <div>
      <input
        type="color"
        value="#808000"
      />
      <button
        class="tag-cloud_settings_remove-color-btn"
      >
        Remove Color
      </button>
    </div>
    <div>
      <input
        type="color"
        value="#800080"
      />
      <button
        class="tag-cloud_settings_remove-color-btn"
      >
        Remove Color
      </button>
    </div>
    <div>
      <input
        type="color"
        value="#C0C0C0"
      />
      <button
        class="tag-cloud_settings_remove-color-btn"
      >
        Remove Color
      </button>
    </div>
    <div>
      <input
        type="color"
        value="#008080"
      />
      <button
        class="tag-cloud_settings_remove-color-btn"
      >
        Remove Color
      </button>
    </div>
    <div>
      <input
        type="color"
        value="#FF6347"
      />
      <button
        class="tag-cloud_settings_remove-color-btn"
      >
        Remove Color
      </button>
    </div>
    <button
      class="tag-cloud_settings_add-color-btn"
    >
      Add Color
    </button>
  </div>
</div>
`;
